x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -588,6 +620,76 @@ private LogicalPlan resolveLookup(Lookup l, List<Attribute> childrenOutput) {
             return l;
         }
 
+        private Join resolveLookupJoin(LookupJoin join) {
+            JoinConfig config = join.config();
+            // for now, support only (LEFT) USING clauses
+            JoinType type = config.type();
+            // rewrite the join into a equi-join between the field with the same name between left and right
+            // per SQL standard, the USING columns are placed first in the output, followed by the rest of left, then right
+            if (type instanceof UsingJoinType using) {
+                List<Attribute> cols = using.columns();
+                // the lookup cannot be resolved, bail out
+                if (Expressions.anyMatch(cols, c -> c instanceof UnresolvedAttribute ua && ua.customMessage())) {
+                    return join;
+                }
+
+                JoinType coreJoin = using.coreJoin();
+                // verify the join type
+                if (coreJoin != JoinTypes.LEFT) {
+                    String name = cols.get(0).name();
+                    UnresolvedAttribute errorAttribute = new UnresolvedAttribute(
+                        join.source(),
+                        name,
+                        "Only LEFT join is supported with USING"
+                    );
+                    return join.withConfig(new JoinConfig(type, singletonList(errorAttribute), emptyList(), emptyList()));
+                }
+                // resolve the using columns against the left and the right side then assemble the new join config
+                List<Attribute> leftKeys = resolveUsingColumns(cols, join.left().output(), "left");
+                List<Attribute> rightKeys = resolveUsingColumns(cols, join.right().output(), "right");
+                List<Attribute> output = new ArrayList<>(join.left().output());
+                output.addAll(join.right().outputSet().subtract(new AttributeSet(rightKeys)));
```

Comment:
Is it really correct to subtract the right hand side's attributes based on name id, which we do here?

I'd expect the left and right hand side's attributes to _normally_ have completely disjoint output sets, esp. because we normally plan to join different indices. They _should_ not have field attributes with the same name ids; instead, we should match them up based on their names, no?

Reply:
See #116208 "clarify output order" - tl;dr this is not final.
Longer:
you are right that each side, including common keys have different ids. the goal is to move away from name matching and convert the using into a equality condition which can then be pushed down and combined with per-side filtering.
That is, formalize the using into an equal-join that is a filter with equality filter which simplifies other optimizations such as transitive filtering (left_key < 1 AND left_key == right_key --> implies right_key < 1; the left_key filter and right_key filter end up pushed down on their respective sides).
This means moving away from JoinCondition to an expression